Add complex support for torch.addcmul and torch.addcdiv

anjali411 opened the pull request:
Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46639 Add complex support for torch.addcmul and torch.addcdiv**

Differential Revision: [D24879099](https://our.internmc.facebook.com/intern/diff/D24879099)

anjali411 committed:
Add complex support for torch.addcmul and torch.addcdiv

[ghstack-poisoned]

anjali411 commented on tools/autograd/derivatives.yaml:
```diff
@@ -177,14 +177,14 @@
-  self: grad
-  tensor1: grad * tensor2 * value
-  tensor2: grad * tensor1 * value
+  self: handle_r_to_c(self.scalar_type(), grad)
```
`addcmul` and `addcdiv` currently don't support type promotion rn so `R -> C` case would never occur but adding this as a precaution in case the type promotion is enabled for these ops in future.

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245



[ghstack-poisoned]

zou3519 commented on test/test_torch.py:
```diff
@@ -12513,10 +12513,6 @@ def rand_tensor(size, dtype, device):
-                continue
-
             actual = torch.addcmul(a, b, c, value=alpha)
             expected = a + alpha * b * c
```
I'm confused -- why isn't there a `self.assertEqual(actual, expected)` check here?

zou3519 replied:
nvm, it looks like github is having problems showing the code: it skips straight from like 12521 to 13436

zou3519 commented on test/test_torch.py:
```diff
@@ -13436,7 +13432,7 @@ def test_addcdiv(self, device):
             actual = torch.addcdiv(a, b, c, value=alpha)
             # implementation of addcdiv downcasts alpha. arithmetic ops don't.
-            if not actual.dtype.is_floating_point:
+            if not (actual.dtype.is_floating_point or actual.dtype.is_complex):
```
I don't see a test for if `alpha` is complex for addcdiv or addcmul, that would be good to add for completeness.

anjali411 replied:
ha I didn't notice that thanks! updated the tests

zou3519 commented on test/test_torch.py:
```diff
@@ -13436,7 +13432,7 @@ def test_addcdiv(self, device):
             # implementation of addcdiv downcasts alpha. arithmetic ops don't.
-            if not actual.dtype.is_floating_point:
+            if not (actual.dtype.is_floating_point or actual.dtype.is_complex):
                 alpha = int(alpha)
```
Github is having a hard time loading test_torch. However, the int(alpha) call looks a bit fishy to me, because the helper function calls _test_addcdiv with 0.5. `int(0.5)` resolves to 0, so the choice of 0.5 is not the best thing for testing integer dtypes with addcdiv. But this is a preexisting problem.
```
def _helper():
    _test_addcdiv(
        non_zero_rand((2, 2), dtype=dtype, device=device),
        0.5,
        non_zero_rand((2, 2), dtype=dtype, device=device),
        non_zero_rand((2, 2), dtype=dtype, device=device))
```

anjali411 replied:
agreed updated!

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245



[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245



[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245



[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245

Differential Revision: [D24879099](https://our.internmc.facebook.com/intern/diff/D24879099)

[ghstack-poisoned]

zou3519 commented on test/test_autograd.py:
```diff
@@ -5006,7 +5006,7 @@ def run_functional_checks(test_case, test_name, name, apply_fn, run_grad_checks,
                 'cosh', '__rmul__', 'sgn', 'abs', 'dot', 'vdot', 'tensor_split', 'matmul',
                 'bmm', 'mv', 'ger', 'diagonal', 'atan', 'angle', 'tanh', 'fill_', 'sub',
-                'exp', 'mean', 'addcmul', 'addcdiv'] + separate_complex_tests
+                'exp', 'mean', 'inverse', 'addcmul', 'addcdiv'] + separate_complex_tests
```
Maybe this should just be a list to avoid merge conflicts lol:
```
complex_list = [
    't',
    'view',
    'reshape',
    'reshape_as',
    ...
]
```
The downside of this is that when you're done adding operators this list is going to be 600 lines long

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245

Differential Revision: [D24879099](https://our.internmc.facebook.com/intern/diff/D24879099)

[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245

Differential Revision: [D24879099](https://our.internmc.facebook.com/intern/diff/D24879099)

[ghstack-poisoned]

anjali411 committed:
Update on "Add complex support for torch.addcmul and torch.addcdiv"

Resolves: https://github.com/pytorch/pytorch/issues/46546#issuecomment-713122245

Differential Revision: [D24879099](https://our.internmc.facebook.com/intern/diff/D24879099)

[ghstack-poisoned]
